Mathlib/Probability/Kernel/Disintegration/CdfToKernel.lean

```diff
@@ -0,0 +1,690 @@
+
+section IsCondKernelCDF
+
+variable {f : α × β → StieltjesFunction}
+
+/-- A function `f : α × β → StieltjesFunction` is called a conditional kernel CDF of `κ` with
+respect to `ν` if it is measurable, tends to to 0 at -∞ and to 1 at +∞ for all `p : α × β`,
+if `fun b ↦ f (a, b) x` is `(ν a)`-integrable for all `a : α` and `x : ℝ` and for all
+measurable sets `s : Set β`, `∫ b in s, f (a, b) x ∂(ν a) = (κ a (s ×ˢ Iic x)).toReal`. -/
+structure IsCondKernelCDF (f : α × β → StieltjesFunction) (κ : kernel α (β × ℝ)) (ν : kernel α β) :
+    Prop :=
+  (measurable (x : ℝ) : Measurable fun p ↦ f p x)
+  (integrable (a : α) (x : ℝ) : Integrable (fun b ↦ f (a, b) x) (ν a))
+  (tendsto_atTop_one (p : α × β) : Tendsto (f p) atTop (𝓝 1))
+  (tendsto_atBot_zero (p : α × β) : Tendsto (f p) atBot (𝓝 0))
+  (set_integral (a : α) {s : Set β} (_hs : MeasurableSet s) (x : ℝ) :
+    ∫ b in s, f (a, b) x ∂(ν a) = (κ a (s ×ˢ Iic x)).toReal)
+
+lemma IsCondKernelCDF.nonneg (hf : IsCondKernelCDF f κ ν) (p : α × β) (x : ℝ) : 0 ≤ f p x :=
+  Monotone.le_of_tendsto (f p).mono (hf.tendsto_atBot_zero p) x
+
+lemma IsCondKernelCDF.le_one (hf : IsCondKernelCDF f κ ν) (p : α × β) (x : ℝ) : f p x ≤ 1 :=
+  Monotone.ge_of_tendsto (f p).mono (hf.tendsto_atTop_one p) x
+
+lemma IsCondKernelCDF.integral
+    {f : α × β → StieltjesFunction} (hf : IsCondKernelCDF f κ ν) (a : α) (x : ℝ) :
+    ∫ b, f (a, b) x ∂(ν a) = (κ a (univ ×ˢ Iic x)).toReal := by
+  rw [← hf.set_integral _ MeasurableSet.univ, Measure.restrict_univ]
+
+lemma IsCondKernelCDF.set_lintegral [IsFiniteKernel κ]
+    {f : α × β → StieltjesFunction} (hf : IsCondKernelCDF f κ ν)
+    (a : α) {s : Set β} (hs : MeasurableSet s) (x : ℝ) :
+    ∫⁻ b in s, ENNReal.ofReal (f (a, b) x) ∂(ν a) = κ a (s ×ˢ Iic x) := by
+  rw [← ofReal_integral_eq_lintegral_ofReal (hf.integrable a x).restrict
+    (ae_of_all _ (fun _ ↦ hf.nonneg _ _)), hf.set_integral a hs x, ENNReal.ofReal_toReal]
+  exact measure_ne_top _ _
+
+lemma IsCondKernelCDF.lintegral [IsFiniteKernel κ]
+    {f : α × β → StieltjesFunction} (hf : IsCondKernelCDF f κ ν) (a : α) (x : ℝ) :
+    ∫⁻ b, ENNReal.ofReal (f (a, b) x) ∂(ν a) = κ a (univ ×ˢ Iic x) := by
+  rw [← hf.set_lintegral _ MeasurableSet.univ, Measure.restrict_univ]
+
+lemma isCondKernelCDF_stieltjesOfMeasurableRat {f : α × β → ℚ → ℝ} (hf : IsRatCondKernelCDF f κ ν)
+    [IsFiniteKernel κ] :
+    IsCondKernelCDF (stieltjesOfMeasurableRat f hf.measurable) κ ν where
+  measurable := measurable_stieltjesOfMeasurableRat hf.measurable
+  integrable := integrable_stieltjesOfMeasurableRat hf
+  tendsto_atTop_one := tendsto_stieltjesOfMeasurableRat_atTop hf.measurable
+  tendsto_atBot_zero := tendsto_stieltjesOfMeasurableRat_atBot hf.measurable
+  set_integral a _ hs x := set_integral_stieltjesOfMeasurableRat hf a x hs
+
+end IsCondKernelCDF
+
+section ToKernel
+
+variable {_ : MeasurableSpace β} {f : α × β → StieltjesFunction}
+  {κ : kernel α (β × ℝ)} {ν : kernel α β} {hf : IsCondKernelCDF f κ ν}
```

Comment:
Could you add `hf` to the theorems that need it? Having `Prop` variables around makes the theorems much harder to parse.